### PR TITLE
rgw: Small refactor and two bug fixes

### PR DESCRIPTION
**Pull request by adamemerson**
Refactor something rather than leaving it open-coded and take the opportunity to fix evaluation behavior in a few cases.

Also fix the DeleteBucketWebiste bug and pre-emptively fix the few cognate cases that hadn't been caught yet.

### CODE REVIEW

**Comment by adamemerson** [comment]
@C2python Does this do what you were hoping for?

**Comment by C2python** [comment]
Yes, this looks better.

File: src/rgw/rgw_common.cc [hunk 1]
```diff
@@ -1152,18 +1153,22 @@ bool verify_bucket_permission(struct req_state * const s, const uint64_t op)
...
-  if (s->iam_policy) {
-    if (s->iam_policy->eval(s->env, *s->auth.identity, op,
-			    ARN(s->bucket)) == Effect::Allow) {
-      return 0;
-    }
-  } else if (s->auth.identity->is_owner_of(s->bucket_owner.get_id())) {
-    return 0;
-  }
-  return -EACCES;
+  auto e = maybe_do_or(s->iam_policy,
```
Comment (mattbenjamin) [inline src/rgw/rgw_common.cc]: I like this idea in theory, but in this particular case, the functional abstraction seems a lot less explicit, and I'm kind of suspecting, maybe adds a bit of overhead?

**Comment by adamemerson** [comment]
@mattbenjamin How's this? Passes s3-tests.

**Comment by adamemerson** [comment]
Jenkins, retest this please.

**Comment by yuriw** [comment]
[wip-yuri2-testing-2017-11-01-1720](https://trello.com/c/jcK1JJvd/192-wip-yuri2-testing-2017-11-01-1720)

**Pull request merged**